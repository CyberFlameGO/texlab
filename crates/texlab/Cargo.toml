[package]
name = "texlab"
description = "LaTeX Language Server"
version = "5.5.1"
license.workspace = true
readme = "README.md"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
homepage = "https://github.com/latex-lsp/texlab"
repository = "https://github.com/latex-lsp/texlab"
documentation = "https://github.com/latex-lsp/texlab"
keywords = ["lsp", "server", "latex", "bibtex"]
categories = ["development-tools"]
exclude = [
  ".gitattributes",
  ".gitignore",
  ".github/**",
  "tests/it/**",
  "*.snap",
  "texlab.1",
  "texlab.pdf",
  "texlab.tex",
]

[[bin]]
name = "texlab"
test = false

[lib]
doctest = false

[dependencies]
anyhow = "1.0.70"
base-db = { path = "../base-db" }
citeproc = { path = "../citeproc" }
clap = { version = "4.2.2", features = ["derive"] }
commands = { path = "../commands" }
crossbeam-channel = "0.5.8"
dirs = "5.0.0"
distro = { path = "../distro" }
encoding_rs = "0.8.32"
encoding_rs_io = "0.1.7"
fern = "0.6.2"
flate2 = "1.0.25"
fuzzy-matcher = { version = "0.3.7", features = ["compact"] }
itertools = "0.10.5"
log = "0.4.17"
lsp-server = "0.7.0"
lsp-types = "0.94.0"
notify = "5.1.0"
once_cell = "1.17.1"
parking_lot = "0.12.1"
parser = { path = "../parser" }
regex = "1.7.3"
rowan = "0.15.11"
rustc-hash = "1.1.0"
serde = "1.0.160"
serde_json = "1.0.96"
serde_regex = "1.1.0"
serde_repr = "0.1.12"
smol_str = { version = "0.2.0", features = ["serde"] }
syntax = { path = "../syntax" }
tempfile = "3.5.0"
threadpool = "1.8.1"
titlecase = "2.2.1"
symbols = { path = "../symbols" }

[dev-dependencies]
assert_unordered = "0.3.5"
criterion = { version = "0.4.0" }
insta = { version = "1.28.0", features = ["glob", "redactions", "json"] }

[[bench]]
name = "bench_main"
path = "benches/bench_main.rs"
harness = false
