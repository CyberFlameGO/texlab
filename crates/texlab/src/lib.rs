mod client;
pub mod features;
mod server;
pub mod util;

pub use self::{client::LspClient, server::Server};

pub(crate) fn normalize_uri(uri: &mut lsp_types::Url) {
    fn fix_drive_letter(text: &str) -> Option<String> {
        if !text.is_ascii() {
            return None;
        }

        match &text[1..] {
            ":" => Some(text.to_ascii_uppercase()),
            "%3A" | "%3a" => Some(format!("{}:", text[0..1].to_ascii_uppercase())),
            _ => None,
        }
    }

    if let Some(mut segments) = uri.path_segments() {
        if let Some(mut path) = segments.next().and_then(fix_drive_letter) {
            for segment in segments {
                path.push('/');
                path.push_str(segment);
            }

            uri.set_path(&path);
        }
    }

    uri.set_fragment(None);
}
