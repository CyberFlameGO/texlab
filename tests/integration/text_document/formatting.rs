use anyhow::Result;
use insta::assert_snapshot;
use lsp_types::{
    request::Formatting, ClientCapabilities, DocumentFormattingParams, FormattingOptions,
    TextDocumentIdentifier,
};
use texlab::{LineIndex, LineIndexExt};

use crate::{client::Client, fixture};

fn format(fixture: &str) -> Result<String> {
    let mut client = Client::spawn()?;
    client.initialize(ClientCapabilities::default(), None)?;

    let fixture = fixture::parse(fixture);
    let file = fixture.files.into_iter().next().unwrap();
    let old_text = file.text.clone();
    client.open(file.name, file.lang, file.text)?;

    let actual_edits = client
        .request::<Formatting>(DocumentFormattingParams {
            text_document: TextDocumentIdentifier::new(client.uri(file.name)?),
            work_done_progress_params: Default::default(),
            options: FormattingOptions {
                insert_spaces: true,
                tab_size: 4,
                ..Default::default()
            },
        })?
        .unwrap_or_default();

    client.shutdown()?;

    let line_index = LineIndex::new(&old_text);
    let mut actual_text = old_text;
    for edit in actual_edits.into_iter().rev() {
        let range = line_index.offset_lsp_range(edit.range);
        actual_text.replace_range::<std::ops::Range<usize>>(range.into(), &edit.new_text);
    }

    Ok(actual_text)
}

#[test]
fn bibtex_internal_wrap_long_lines() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = {Lorem ipsum dolor sit amet, consectetur adipiscing elit.
%SRC Lorem ipsum dolor sit amet,
%SRC consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.},}"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_multiple_entries() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = {Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, 
%SRC consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.},}
%SRC 
%SRC @article{foo, bar = {Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, 
%SRC consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.},}""#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_trailing_comma() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = baz}"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_insert_braces() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = baz,"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_command() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = "\baz",}"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_join_strings() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article{foo, bar = "baz" # "qux"}"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_parens() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @article(foo,)"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_string() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @string{foo="bar"}"#,
    )?);

    Ok(())
}

#[test]
fn bibtex_internal_preamble() -> Result<()> {
    assert_snapshot!(format(
        r#"
%BIB main.bib
%SRC @preamble{
%SRC     "foo bar baz" }"#,
    )?);

    Ok(())
}
