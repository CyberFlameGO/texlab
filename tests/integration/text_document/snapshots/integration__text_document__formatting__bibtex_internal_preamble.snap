---
source: tests/integration/text_document/formatting.rs
assertion_line: 140
expression: "format(r#\"\n%BIB main.bib\n%SRC @preamble{\n%SRC     \"foo bar baz\" }\"#)?"
---
@preamble{"foo bar baz"}

